README
Courier::Filter 0.17
A purely Perl-based mail filter framework for the Courier MTA.
(C) 2003-2005 Julian Mehnle <julian@mehnle.net>
<http://www.mehnle.net/software/courier-filter>
==============================================================================

The Courier MTA's courierfilter interface
-----------------------------------------

The Courier MTA <http://www.courier-mta.org> offers an interface for
daemon-style processes to act as mail filters, called courierfilters.  For
every incoming mail message, right after the DATA command in the SMTP
transaction phase has completed, Courier calls every registered mail filter
through a UNIX domain socket the filter is listening on, and feeds it the file
names of the incoming message and one or more control files.  The mail filter
processes the message and its control file(s), and returns an SMTP-style
status response.  If the status response is a positive ("2xx") one, Courier
accepts the message.  Otherwise, Courier rejects the message using the
returned response code and text.

About Courier::Filter
---------------------

Courier::Filter implements the courierfilter interface as a framework for mail
filter modules that frees modules from the duties of creating and handling the
UNIX domain sockets, waiting for connections from Courier, and reading and
parsing message and control files.  Thus, authors of filter modules can
concentrate on writing the actual filter logic without having to care about
things that can easily be abstracted and can be performed by the framework.

Courier::Filter allows multiple filter modules to be installed, and filter
modules can be stacked and grouped hierarchically, and even a module's polarity
can be reversed, so some modules can be used for explicitly accepting messages
while others are used in the traditional way for rejecting messages.

Getting started
---------------

Read the following perldocs in order to get started with Courier::Filter:
* Courier::Filter::Overview
* Courier::Filter
* Courier::Filter::Module

Upgrading from versions prior to 0.15
-------------------------------------

Starting from version 0.15, the Courier::* Perl modules are no longer being
installed in your Perl's configured "vendorlib" path, but in the private path
'/usr/share/courier-filter-perl/perl5'.  (This is because Courier::Filter is
primarily a plug-in for the Courier MTA, and not a library of Perl modules that
are intended to be used by other programs or modules.  However some of the
modules (e.g. Courier::Message) are indeed designed genericly, and should any
other program or module actually make use of them in the future, I will split
them off into a separate library package.)

Due to this partial relocation, upgrading from a version prior to 0.15 is
probably best achieved by removing the old version and doing a fresh install.
You can of course keep your old configuration file(s).

License
-------

Courier::Filter is free software.  You may use, modify, and distribute it under
the same terms as Perl itself, i.e. under the GNU GPL (version 2 or later) or
the Artistic License.

# $Id: README 199 2005-11-10 22:16:37Z julian $
# vim:tw=79


CHANGES
# Legend:
# --- = A new release
#   x = Changed something significant, or removed a feature
#   * = Fixed a bug, or made a minor improvement
#   + = Added a feature (in a backwards compatible way)

--- 0.17 (2005-11-11 03:00)

  Debian:
  * Recommends: libmime-perl, not libmime-tools-perl.  (Closes Debian bug
    #297730)
  x Recommends: libio-stringy-perl (>= 2.110), libmail-spamassassin-perl,
    libarchive-zip-perl.

  Courier::Filter:
  * If socket already exists, check whether it is dead or alive.  If it is
    dead, remove it and start up normally.  If it is alive, die like before.
    (Closes Debian bug #326785)
  * Ignore connections that don't send any data.

  Courier::Message:
  * Decode header text as "UTF-8" encoded, not as "utf8" encoded.  Starting
    with Perl 5.8.7, this does make a difference.  "utf8" allows invalid
    Unicode codepoints (such as UTF-16 surrogates), while "UTF-8" (the official
    standard UTF-8) doesn't.  We want to avoid invalid codepoints because they
    tend to cause trouble later.
  * Don't die on invalid characters in encoded-word chunks.  If an encoded-word
    cannot be decoded, retain it undecoded.
  * "authenticated" property: refined parsing of "Received:" header and "AUTH:"
    authentication information.
  + Added "authenticated_user" property.
  * Corrected documentation of "header" property.

  SPFout filter module:
  * Don't be pedantic about what our "public" IPv4 address is vis-à-vis the
    various recipient MXes.  Assume we just have a single public address that
    applies to all recipient MXes.  (Let's see if anyone out there actually has
    a more complex setup...)

  Parts filter module:
  * Requires IO::InnerFile 2.110 or later, Archive::Zip.

  SpamAssassin filter module:
  + Added for SpamAssassin-based spam detection.
  x Requires Mail::SpamAssassin.

  Miscellaneous:
  * Relaxed sender IPv4 address matching in DNSBL, SPF, and SPFout filter
    modules, making the "::ffff:" IPv6 prefix optional and thus supporting
    IPv6-disabled versions of Courier.
  * Courier::Filter::Overview: Updated mention of pythonfilter from version 0.5
    to 0.8.
  * README: Added license statement.
  + INSTALL: Added system requirements and building/installation instructions.
  * Minor documentation fixes.

--- 0.16 (2005-01-17 19:11)

  Debian:
  * Explicitly specify the dependency on Perl 5.8 so upgrades from Woody are
    guaranteed to work smoothly and also to simplify backporting.  This deals
    with Debian bug #289796.
  x Recommend libnet-dns-perl.  Suggest libnet-rblclient-perl,
    libnet-address-ipv4-local-perl, and libclamav-client-perl (all of which are
    not yet packaged for Debian).

  SPF filter module:
  x Deprecated the "reject_on" option in favor of the new "match_on" option.
  x Removed "none" from the default set of result codes to match on.
  x Changed the documentation to recommend also rejecting on "unknown" results.
  + Added a reference in the documentation to the spf-draft-200406
    specification, to which this module complies.

  SPFout filter module:
  + Added for outbound SPF filtering.

  ClamAVd filter module:
  + Added for malware filtering using a ClamAV `clamd` dameon.

  Miscellaneous:
  * Updated the sample configuration file.
  * Minor documentation fixes.

--- 0.15 (2004-10-30 01:13)

  x Courier::Filter is primarily a plug-in for the Courier MTA, not a library
    of Perl modules intended to be used by other programs or modules.
    x Renamed the Debian package to "courier-filter-perl".
    x Install included Perl modules in the private path
      /usr/share/courier-filter-perl/perl5 instead of Perl's configured
      "vendorlib" path (/usr/share/perl5 in Debian).
  * Various minor fixes.

--- 0.14 (2004-10-22 00:10)

  + Added documentation for the test-filter-module and pureperlfilter
    executables.
  * test-filter-module:
    * Fixed the command-line parsing and handling of control file names.
    * Made the result output format compatible with multi-line results.

--- 0.13 (2004-10-04 23:00)

  + Added Debian package control files.

  x Deprecated the MIMEParts filter module in favor of the new Parts filter
    module, which is compatible but a lot more powerful:
    x Switched from completely-in-memory processing to using temporary files in
      order to work around two nasty bugs in MIME::Parser and Archive::Zip.
    x Renamed the "max_size" constructor option to "max_message_size".
      "max_size" is deprecated but still supported for now.
    x The "max_size" constructor option now defaults to 1024**2 (1MB) instead
      of undef.
    + Added the "max_part_size" constructor option.
    + Added "views" support (as constructor option and signature aspect),
      supporting "raw" and "zip" views for now.
    + Added the "encrypted" signature aspect.

  + Added Envelope filter module.
  + Added etc/pureperlfilter.conf as a default configuration file.

  * Fixed the mysterious ``Can't call method "close" on an undefined value at
    Filter.pm line 299´´ bug that caused Courier::Filter to fail starting up on
    some platforms.  (File descriptor 3 must be opened in write mode, not in
    read mode!)
  * Fixed filter module group handling:
    * Skip the rest of the current module group and treat it as an implicit
      accept if a module in the group states an explicit accept.
    * Log rejections just once, right after a module has stated a reject,
      i.e. avoid logging the rejection for all higher recursion levels of the
      module group structure that contains the rejecting module.
    * Don't try extracting a logger from filter module groups, only from atomic
      filter modules.
  * Fixed the Header filter module:
    * Fixed handling of undefined header fields.
    * Prepend a "Header:" keyword to positive match results.
  * Disabled timestamping for the Syslog logger module, as syslog does its own
    timestamping.
  * Reworked configuration file example.

lib/Courier/Filter/Module/Parts.pm
#
# Courier::Filter::Module::Parts class
#
# (C) 2003-2005 Julian Mehnle <julian@mehnle.net>
# $Id: Parts.pm 195 2005-06-11 19:47:59Z julian $
#
##############################################################################

=head1 NAME

Courier::Filter::Module::Parts - A message (MIME multipart and ZIP archive)
parts filter module for the Courier::Filter framework

=cut

package Courier::Filter::Module::Parts;

=head1 VERSION

0.17

=cut

our $VERSION = '0.17';

=head1 SYNOPSIS

    use Courier::Filter::Module::Parts;
    
    my $module = Courier::Filter::Module::Parts->new(
        max_message_size
                        => $max_message_size,
        max_part_size   => $max_part_size,
        views           => ['raw', 'zip'],
        signatures      => [
            {
                # One or more of the following options:
                mime_type       => 'text/html' || qr/html/i,
                file_name       => 'file_name.ext' || qr/\.(com|exe)$/i,
                size            => 106496,
                digest_md5      => 'b09e26c292759d654633d3c8ed00d18d',
                encrypted       => 0,
                
                # Optionally any of the following:
                views           => ['raw', 'zip'],
                response        => $response_text
            },
            ...
        ],
        
        logger          => $logger,
        inverse         => 0,
        trusting        => 0,
        testing         => 0,
        debugging       => 0
    );
    
    my $filter = Courier::Filter->new(
        ...
        modules         => [ $module ],
        ...
    );

=cut

use warnings;
use strict;

use base qw(Courier::Filter::Module);

use MIME::Parser 5.4;
use IO::InnerFile 2.110;
    # Require either MIME::Parser 5.413 or lower, or IO::InnerFile 2.110+
    # (where IO::InnerFile::seek() properly returns TRUE when appropriate).
use Digest::MD5;
use File::Spec;
    # In-memory processing doesn't work, see comments in match_mime_part().

# Constants:
##############################################################################

use constant TRUE   => (0 == 0);
use constant FALSE  => not TRUE;

use constant DEFAULT_RESPONSE   => 'Prohibited message part detected.';

# Interface:
##############################################################################

=head1 DESCRIPTION

This class is a filter module class for use with Courier::Filter.  It matches a
message if one of the message's parts (MIME parts, or files in a ZIP archive)
matches one of the configured signatures.

=cut

sub new;

sub match;

# Implementation:
##############################################################################

=head2 Constructor

The following constructor is provided:

=over

=item B<new(%options)>: RETURNS Courier::Filter::Module::Parts

Creates a new B<Parts> filter module.

%options is a list of key/value pairs representing any of the following
options:

=over

=item B<views>

An arrayref containing the global default set of I<views> the filter module
should apply to message parts when matching the configured signatures against
them.  A view is the way how a MIME part's (MIME-decoded) data is interpreted.
Defaults to B<['raw']>.

The following views are supported:

=over

=item B<raw>

The MIME part is MIME-decoded but not otherwise transformed.  The raw MIME part
is then matched against the configured signatures.

=item B<zip>

If the MIME part has a file name ending in C<.zip>, it is considered a ZIP
archive, and all unencrypted files in the archive are matched as individual
message parts against the configured signatures.  The zip view requires the
B<Archive::Zip> Perl module to be installed.

=back

=item B<max_message_size>

=item B<max_size> (DEPRECATED)

An integer value controlling the maximum size (in bytes) of the overall message
text for a message to be processed by this filter module.  Messages larger than
this value will never be processed, and thus will never match.  If B<undef>,
there is no size limit.  Defaults to B<1024**2> (1MB).

As MIME multipart and ZIP archive processing can be quite CPU- and
memory-intensive (although the B<Parts> filter module makes use of temporary
files since version 0.13), you should definitely restrict the message size to
some sensible value that easily fits in your server's memory.  1024**2 (1MB)
should be appropriate for most uses of this filter module.

The C<max_message_size> option was previously called C<max_size>, but the
latter is now deprecated and may not be supported in future versions of the
B<Parts> filter module.

=item B<max_part_size>

An integer value controlling the maximum size (in bytes) of any single message
part (i.e. MIME part in a message, or file in an archive) for that part to be
processed by this filter module.  Parts larger than this value will never be
processed, and thus will never match.  If B<undef>, there is no size limit.

Defaults to the value of the C<max_message_size> option, so you don't really
need to specify a part size limit if you are comfortable with using the same
value for both.  See the C<max_message_size> option for its default.

If you make use of the B<'zip'> view, be aware of the risk posed by so-called
I<decompression bombs>, which allow messages to easily fall below the overall
message size limit, while a file in a small attached ZIP archive can decompress
to a huge size.  The part size limit prevents huge files from being
decompressed.

=item B<signatures>

REQUIRED.  A reference to an array containing the list of I<signatures> against
which message parts are to be matched.  A signature in turn is a reference to a
hash containing one or more so-called signature I<aspects> (as key/value pairs)
and any signature I<options> (also as key/value pairs).

I<Signature aspects>

Aspects may either be scalar values (for exact, case-sensitive matches), or
regular expression objects created with the C<qr//> operator (for inexact,
partial matches).  For a signature to match a message part, I<all> of the
signature's specified aspects must match those of the message part.  For the
filter module to match a message, I<any> of the signatures must match I<any> of
the message's parts.

A signature aspect can be any of the following:

=over

=item B<mime_type>

The MIME type of the message part ('type/sub-type').

=item B<file_name>

The file name of the message part.

=item B<size>

The exact size (in bytes) of the decoded message part.

=item B<digest_md5>

The MD5 digest of the decoded message part (32 hex digits, as printed by
`md5sum`).

=item B<encrypted>

A boolean value denoting whether the message part is encrypted and its contents
are inaccessible to the B<Parts> filter module.

=back

I<Signature options>

A signature option can be any of the following:

=over

=item B<views>

An arrayref containing the set of I<views> the filter module should apply to
message parts when matching I<this> signature against them.  For a list of
supported views, see the description of the constructor's C<views> option.
Defaults to the global set of views specified to the constructor.

=item B<response>

A string that is to be returned as the match result in case of a match.
Defaults to B<"Prohibited message part detected.">.

=back

I<Example>

So for instance, a signature list could look like this:

    signatures  => [
        {
            mime_type   => qr/html/i,
            response    => 'No HTML mail, please.'
        },
        {
            file_name   => qr/\.(com|exe|lnk|pif|scr|vbs)$/i,
            response    => 'Executable content detected'
        },
        {
            size        => 106496,
            digest_md5  => 'b09e26c292759d654633d3c8ed00d18d',
            views       => ['raw', 'zip'],  # Look into ZIP archives, too!
            response    => 'Worm detected: W32.Swen'
        },
        {
            size        => 22528,
            # Cannot set a specific digest_md5 since W32.Mydoom
            # is polymorphic.
            response    => 'Worm suspected: W32.Mydoom'
        },
        {
            encrypted   => 1,
            views       => ['zip'],
            response    => 'Worm suspected ' .
                           '(only worms and fools use ZIP encryption)'
        }
    ]

=back

All options of the B<Courier::Filter::Module> constructor are also supported
by the constructor of the B<Parts> filter module.  Please see
L<Courier::Filter::Module/"new"> for their descriptions.

=cut

sub new {
    my ($class, %options) = @_;
    
    my $mime_parser = MIME::Parser->new();
    #$mime_parser->output_to_core(TRUE);
        # In-memory processing doesn't work, see comments in match_mime_part().
    $mime_parser->output_under(File::Spec->tmpdir);
    #$mime_parser->tmp_to_core(TRUE);
        # In-memory processing doesn't work, see comments in match_mime_part().
    $mime_parser->use_inner_files(TRUE);
    
    my $module = $class->SUPER::new(
        %options,
        mime_parser => $mime_parser
    );
    
    # Default "max_message_size" option to the deprecated "max_size" option,
    # or to 1024**2 (1MB):
    $module->{max_message_size} = (
        exists($module->{max_size}) ? $module->{max_size} : 1024**2
    )
        if not exists($module->{max_message_size});
    
    # Default "max_part_size" option to the "max_message_size" option:
    $module->{max_part_size} = $module->{max_message_size}
        if not exists($module->{max_part_size});
    
    # Default "views" option to 'raw':
    my $views = $module->{views} || { 'raw' => TRUE };
    
    # Transform "views" option into hashref if it was given as an arrayref:
    $views = { map(($_ => TRUE), @$views) }
        if ref($views) eq 'ARRAY';
    
    my $used_views = { %$views };
    foreach my $signature ( @{$module->{signatures}} ) {
        # Default "views" option to global "views" option:
        my $signature_views = $signature->{views} || $views;
        
        # Transform "views" option into hashref if it was given as an arrayref:
        $signature_views = { map(($_ => TRUE), @$signature_views) }
            if ref($signature_views) eq 'ARRAY';
        
        # Add any signature-specific views to the global set of used views:
        %$used_views = (%$used_views, %$signature_views);

        $signature->{views} = $signature_views;
        
        $module->compile_signature($signature);
    }

    $module->{used_views} = $used_views;
    
    return $module;
}

=back

=head2 Instance methods

See L<Courier::Filter::Module/"Instance methods"> for a description of the
provided instance methods.

=cut

sub match {
    my ($module, $message) = @_;
    my $class = ref($module);
    
    return undef
	if  defined($module->{max_message_size})
	and -s $message->file_name > $module->{max_message_size};
    
    #my $text = $message->text;
    #my $part = $module->{mime_parser}->parse_data($text);
        # In-memory processing doesn't work, see comments in match_mime_part().
    my $part = $module->{mime_parser}->parse_open($message->file_name);
    my ($result, @code) = $module->match_mime_part($part);
    
    $result &&= 'Parts: ' . $result;
    
    $module->{mime_parser}->filer->purge();
        # In-memory processing doesn't work, see comments in match_mime_part().
    rmdir($module->{mime_parser}->filer->output_dir);
        #if MIME::Tools->VERSION < 6.0;
        # Purging also doesn't work properly
        # (bug filed: <http://rt.cpan.org/NoAuth/Bug.html?id=7858>).
    
    return ($result, @code);
}

sub match_mime_part {
    my ($module, $part) = @_;
    
    if (my $body = $part->bodyhandle) {
        # No sub-parts, match this part.
        
        #my $handle = $body->open('r');
            # In-memory processing doesn't work because MIME::Body::open()
            # doesn't provide a fully-IO::Handle-compatible I/O handle object
            # (opened() method is missing, no bug filed).  Working around that
            # by alternatively creating a Perl 5.8 style in-memory file
            # object...
            #   my $body_as_string = $body->as_string;
            #   open(my $handle, '+<', \$body_as_string);
            # ...doesn't work either because Archive::Zip::_isSeekable() is
            # broken (erroneously considers Perl 5.8 style in-memory IO::File
            # objects not to be seekable,
            # bug filed: <http://rt.cpan.org/NoAuth/Bug.html?id=7855>).
            # All of this forces us to make MIME::Parser use temporary files
            # instead of doing everything exclusively in-memory.  Aaargh!!
	
        # First, we gather signature makers for all possible (and enabled)
        # views of the MIME part, then we actually test each view in turn
        # against the configured test signatures.
        
        my @views;
        
        # Raw view (the MIME part itself) (if enabled):
        my $rawsig = $module->make_signature_from_mime_part($part);
        if ($module->{used_views}->{'raw'}) {
            push(
                @views,
                {
                    name        => 'raw',
                    sig_maker   => sub { $rawsig }
                }
            )
                if not defined($module->{max_part_size})
                or $rawsig->{size} <= $module->{max_part_size};
        }
        
        # ZIP archive members view (if enabled and MIME part is a ZIP archive):
        if (
            $module->{used_views}->{'zip'} and
            defined($rawsig->{file_name}) and
            $rawsig->{file_name} =~ /\.zip$/i
        ) {
            require Archive::Zip;
            
            my $archive = Archive::Zip->new();
            #$archive->readFromFileHandle($handle);
                # In-memory processing doesn't work, see above.
            $archive->read($body->path);
            
            # Make a view for each archive member:
            foreach my $member ($archive->members) {
                push(
                    @views,
                    {
                        name        => 'zip',
                        sig_maker   => sub {
                            $module->make_signature_from_zip_archive_member($member)
                        }
                    }
                )
                    if not defined($module->{max_part_size})
                    or $member->uncompressedSize <= $module->{max_part_size};
            }
        }
        
        # Now, for each view, try matching the configured signatures:
        foreach my $view (@views) {
            # Make signature from data view:
            my $datasig = $view->{sig_maker}->();

            # Test that signature against the configured signatures:
            foreach my $signature ( @{$module->{signatures}} ) {
                # Skip this signature if it doesn't apply to the current view:
                next if not $signature->{views}->{ $view->{name} };
                
                my ($result, @code) = $signature->{matcher}->($datasig);
                return ($result, @code) if defined($result);
            }
        }
    }
    else {
	# Match all sub-parts:
	foreach my $subpart ($part->parts) {
	    my ($result, @code) = $module->match_mime_part($subpart);
	    return ($result, @code) if defined($result);
	}
    }
    
    return undef;
}

sub make_signature_from_mime_part {
    my ($module, $part) = @_;
    
    my $head = $part->head;
    my $body = $part->bodyhandle;
    my $text = $body->as_string;
    
    return {
        mime_type   => $head->mime_type,
        file_name   => $head->recommended_filename,
        size        => length($text),
        digest_md5  => Digest::MD5::md5_hex($text),
        encrypted   => FALSE
    };
}

sub make_signature_from_zip_archive_member {
    my ($module, $member) = @_;
    
    return {
        mime_type   => undef,
        file_name   => $member->fileName,
        size        => $member->uncompressedSize,
        digest_md5  => $member->isEncrypted ?
                           undef
                       :   Digest::MD5::md5_hex(scalar($member->contents)),
        encrypted   => $member->isEncrypted
    };
}

sub compile_signature {
    my ($module, $signature) = @_;

    my %matchers;

    my @aspects = grep(!/^(?:response|views)$/, keys(%$signature));
    foreach my $aspect (@aspects) {
        my $pattern = $signature->{$aspect};
        
        my $matcher;
        if (ref($pattern) eq 'Regexp') {
            $matcher = sub { $_[0] =~ $pattern };
        }
        elsif (ref($pattern) eq 'CODE') {
            $matcher = $pattern;
        }
        else {
            if ($aspect =~ /^(?:encrypted)$/) {
                # Aspect is of boolean type:
                $matcher = sub { not ($_[0] xor $pattern) };
            }
            else {
                $matcher = sub { $_[0] eq $pattern };
            }
        }
        
        $matchers{$aspect} = $matcher;
    }
    
    my @response =
        ref($signature->{response}) eq 'ARRAY' ?
            @{ $signature->{response} }
        :   ($signature->{response} || DEFAULT_RESPONSE);
    
    my $matcher = sub {
        # Closure with regard to %matchers.
        my ($signature) = @_;
        foreach my $aspect (keys(%matchers)) {
            my $value = $signature->{$aspect};
            return undef
                if not defined($value)
                or not $matchers{$aspect}->($value);
        }
        return @response;
    };
    
    $signature->{matcher} = $matcher;
    
    return;
}

=head1 SEE ALSO

L<Courier::Filter::Module>, L<Courier::Filter::Overview>.

For AVAILABILITY, SUPPORT, and LICENSE information, see
L<Courier::Filter::Overview>.

=head1 AUTHOR

Julian Mehnle <julian@mehnle.net>

=cut

TRUE;

# vim:tw=79
